[FSDP] Add re-key btw param names/IDs for optim state dict

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#74912 [FSDP] Add re-key btw param names/IDs for optim state dict**
* #74879 [FSDP] Optim state chkpt: key by param name, not ID
* #74215 [FSDP] Add full optim state dict
* #74834 [Easy][FSDP] (Reland) Doc fixes

**Overview**
This introduces a new static method `FSDP.rekey_optim_state_dict()` as a utility for interoperating between local/DDP (non-wrapped) models and FSDP (wrapped) models.

To load from a wrapped model to a non-wrapped model:
```
wrapped_model, wrapped_optim = ...
full_osd = FSDP.full_optim_state_dict(wrapped_model, wrapped_optim)
nonwrapped_model, nonwrapped_optim = ...
rekeyed_osd = FSDP.rekey_optim_state_dict(full_osd, OptimStateKeyType.PARAM_ID, nonwrapped_model)
nonwrapped_optim.load_state_dict(rekeyed_osd)
```
To load from a non-wrapped model to a wrapped model:
```
nonwrapped_model, nonwrapped_optim = ...
osd = nonwrapped_optim.state_dict()
rekeyed_osd = FSDP.rekey_optim_state_dict(osd, OptimStateKeyType.PARAM_NAME, nonwrapped_model)
wrapped_model, wrapped_optim = ...
sharded_osd = FSDP.shard_full_optim_state_dict(rekeyed_osd, wrapped_model)
wrapped_optim.load_state_dict(sharded_osd)
```

**Test Plan**
`test_rekey_optim_state_dict_to_ids()` and `test_rekey_optim_state_dict_to_names()`.

Differential Revision: [D35225819](https://our.internmc.facebook.com/intern/diff/D35225819)

Commit:
[FSDP] Add re-key btw param names/IDs for optim state dict

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -513,14 +511,15 @@ def _get_flat_param_to_fsdp_module(
 
 
-def _get_flat_param_id_to_param(
+def _get_param_id_to_param(
```
Renaming this because it actually works for parameter ID to parameter in both the flattened and unflattened cases, as long as the keys and values are consistent.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2273,6 +2279,107 @@ def shard_full_optim_state_dict(
         }
         return sharded_optim_state_dict
 
+    @staticmethod
```
This functionality is actually not unique to `FSDP`. It _could_ be used generally for PyTorch. However, nowhere else have we seen keying by parameter name, so I have put this inside `fully_sharded_data_parallel.py` for now.

zhaojuanmao replied:
sounds good!

awgu commented on torch/distributed/fsdp/optim_utils.py:
```diff
@@ -346,10 +346,7 @@ def _flatten_tensor_optim_state(
             state).
     """
     non_none_tensors = [t for t in pos_dim_tensors if t is not None]
-    # Check that all are tensors on CPU with the same dtype
```
Just removing this check, which is overly strict. We can just move the tensors to CPU inside the function to avoid device mismatch. It is not actually semantically important which device the tensors in the optimizer state are on.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -189,6 +190,11 @@ class StateDictType(Enum):
     SHARDED_STATE_DICT = auto()
 
 
+class OptimStateKeyType(Enum):
```
nit: let's export it in fspd/__init__.py file as well?

zhaojuanmao reviewed (approved):
looks great!

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225819).

Commit:
Update on "[FSDP] Add re-key btw param names/IDs for optim state dict"


**Overview**
This introduces a new static method `FSDP.rekey_optim_state_dict()` as a utility for interoperating between local/DDP (non-wrapped) models and FSDP (wrapped) models.

To load from a wrapped model to a non-wrapped model:
```
wrapped_model, wrapped_optim = ...
full_osd = FSDP.full_optim_state_dict(wrapped_model, wrapped_optim)
nonwrapped_model, nonwrapped_optim = ...
rekeyed_osd = FSDP.rekey_optim_state_dict(full_osd, OptimStateKeyType.PARAM_ID, nonwrapped_model)
nonwrapped_optim.load_state_dict(rekeyed_osd)
```
To load from a non-wrapped model to a wrapped model:
```
nonwrapped_model, nonwrapped_optim = ...
osd = nonwrapped_optim.state_dict()
rekeyed_osd = FSDP.rekey_optim_state_dict(osd, OptimStateKeyType.PARAM_NAME, nonwrapped_model)
wrapped_model, wrapped_optim = ...
sharded_osd = FSDP.shard_full_optim_state_dict(rekeyed_osd, wrapped_model)
wrapped_optim.load_state_dict(sharded_osd)
```

**Test Plan**
`test_rekey_optim_state_dict_to_ids()` and `test_rekey_optim_state_dict_to_names()`.

Differential Revision: [D35225819](https://our.internmc.facebook.com/intern/diff/D35225819)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225819).
